File: built-in plugins dsp/mb-delay-mod.cpp
/**
 * MB Modulated Delay
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Modulated delay with chorus
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DELAY_MOD_H
#define MB_DELAY_MOD_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDelayMod : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-delay-mod";
    static constexpr const char* PLUGIN_NAME    = "MB Modulated Delay";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float time = 350f;  // range [50, 1500]
    float modRate = 0.5f;  // range [0.1, 5]
    float modDepth = 0.3f;  // range [0, 1]
    float mix = 0.3f;  // range [0, 1]
    };

    MbDelayMod() = default;
    ~MbDelayMod() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.time = std::clamp(params.time, 50f, 1500f);
        params.modRate = std::clamp(params.modRate, 0.1f, 5f);
        params.modDepth = std::clamp(params.modDepth, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Modulated Delay
        return input;
    }
};

#endif // MB_DELAY_MOD_H

File: maxbooster7.5/built-in plugins dsp/mb-delay-mod.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDelayModPlugin: PluginDefinition = {
  id: "mb-delay-mod",
  slug: "mb-delay-mod",
  name: "MB Modulated Delay",
  category: "effect",
  type: "delay",
  version: "1.0.0",
  description: "Modulated delay with chorus",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "time",
      name: "Time",
      type: "float",
      defaultValue: 350,
      minValue: 50,
      maxValue: 1500,
      automatable: true,
    },
    {
      id: "modRate",
      name: "Mod Rate",
      type: "float",
      defaultValue: 0.5,
      minValue: 0.1,
      maxValue: 5,
      automatable: true,
    },
    {
      id: "modDepth",
      name: "Mod Depth",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { time: 350, modRate: 0.5, modDepth: 0.3, mix: 0.3 },
};

export default MbDelayModPlugin;
